modules/crm_tools/crm_tools.module

```diff
@@ -0,0 +1,353 @@
+        '#title' => t('Parent terms'),
+        '#options' => $options,
+        '#default_value' => $role->getThirdPartySetting('crm_tools', 'crm_tools_parent'),
+        '#multiple' => FALSE,
+      ];
+
+      $form['crm_tools_hierarchy']['weight'] = [
+        '#type' => 'textfield',
+        '#title' => t('Weight'),
+        '#size' => 6,
+        '#default_value' => $role->getWeight(),
+        '#description' => t('Roles are displayed in ascending order by weight.'),
+        '#required' => TRUE,
+      ];
+
+      $form['#entity_builders'][] = 'crm_tools_form_user_role_add_form_builder';
+    }
+  }
+}
+
+/**
+ * Entity builder for the menu configuration entity.
+ */
+function crm_tools_form_user_role_add_form_builder($entity_type, Role $role, &$form, FormStateInterface $form_state) {
+  // Update is_hat.
+  if ($form_state->getValue('crm_tools_is_hat')) {
+    $role->setThirdPartySetting('crm_tools', 'crm_tools_is_hat', $form_state->getValue('crm_tools_is_hat'));
+  }
+  else {
+    $role->unsetThirdPartySetting('crm_tools', 'crm_tools_is_hat');
+  }
+
+  // Update hat color.
+  if ($form_state->getValue('color')) {
+    $role->setThirdPartySetting('crm_tools', 'crm_tools_color', $form_state->getValue('color'));
+  }
+  else {
+    $role->unsetThirdPartySetting('crm_tools', 'crm_tools_color');
+  }
+
+  // Update hat icon.
+  if ($form_state->getValue('crm_tools_icon')) {
+    $role->setThirdPartySetting('crm_tools', 'crm_tools_icon', $form_state->getValue('crm_tools_icon'));
+  }
+  else {
+    $role->unsetThirdPartySetting('crm_tools', 'crm_tools_icon');
+  }
+
+  // Update hat parent.
+  if ($form_state->getValue('crm_tools_parent')) {
+    $role->setThirdPartySetting('crm_tools', 'crm_tools_parent', $form_state->getValue('crm_tools_parent'));
+  }
+  else {
+    $role->unsetThirdPartySetting('crm_tools', 'crm_tools_parent');
+  }
+
+  // Update hat weight.
+  if ($form_state->getValue('weight')) {
+    $role->setWeight($form_state->getValue('weight'));
+  }
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_presave() for user entities.
+ */
+function crm_tools_user_presave(UserInterface $account) {
+  $tools = \Drupal::config('crm_tools.tools')->get('active_tools');
+  if (!empty($tools['hats']) && isset($account->original)) {
+    /* @var \Drupal\crm_tools\AdvancedRoleStorageInterface $role_storage */
+    $role_storage = \Drupal::entityTypeManager()->getStorage('user_role');
+    $old_roles = $account->original->getRoles();
+    $new_roles = $account->getRoles();
+
+    // Check that we are adding all parents of new roles.
+    $added_roles = array_diff($new_roles, $old_roles);
+    foreach ($added_roles as $role) {
+      foreach ($role_storage->loadAllParents($role) as $new_role) {
+        if (!$account->hasRole($new_role->id())) {
+          $account->addRole($new_role->id());
+        }
+      }
+    }
+
+    // Check that we are removing all children of old roles.
+    $removed_roles = array_diff($old_roles, $new_roles);
```

Comment:
I think we need to also include any roles added just above. If you add child and remove individual, I think keeping individual makes sense, removing both seems unexpected.